panther_driver/src/panther_can.py

```diff
@@ -172,19 +198,113 @@ def query_runtime_stat_flag(self) -> Generator:
                         )
                         self._error_handle()
                     yield motor_controller.runtime_stat_flag[i]
-        
-    def _error_handle(self) -> None:
-        self._err_times.append(time())
-        self._err_times.pop(0)
 
-    def _turn_on_roboteq_emergency_stop(self) -> None:
+
+class PantherCANUART(PantherCAN):
+    def __init__(self, eds_file, can_interface):
+        super().__init__(eds_file, can_interface)
+
+        while not self._configure_tpdo() and not rospy.is_shutdown():
+            rospy.loginfo(f'[{rospy.get_name()}] Retrying TPDO configuration.')
+            sleep(1)
+
+        self._restart_roboteq_script()
+
+    def query_wheels_enc_pose(self) -> Generator:
         with self._lock:
             for motor_controller in self._motor_controllers:
-                motor_controller.can_node.sdo['Cmd_ESTOP'].raw = 1
+                for pos in motor_controller.wheel_pos:
+                    yield pos
 
-    def _turn_off_roboteq_emergency_stop(self) -> None:
+    def query_wheels_enc_velocity(self) -> Generator:
         with self._lock:
             for motor_controller in self._motor_controllers:
-                motor_controller.can_node.sdo['Cmd_MGO'].raw = 1
+                for vel in motor_controller.wheel_vel:
+                    yield vel
+
+    def query_motor_current(self) -> Generator:
+        with self._lock:
+            for motor_controller in self._motor_controllers:
+                for curr in motor_controller.wheel_curr:
+                    yield curr
+
+    def query_fault_flags(self) -> Generator:
+        with self._lock:
+            for motor_controller in self._motor_controllers:
+                yield motor_controller.fault_flags
+
+    def query_script_flags(self) -> Generator:
+        with self._lock:
+            for motor_controller in self._motor_controllers:
+                yield motor_controller.script_flags
+
+    def query_runtime_stat_flag(self) -> Generator:
+        with self._lock:
+            for motor_controller in self._motor_controllers:
+                for flag in motor_controller.runtime_stat_flag:
+                    yield flag
 
-        
\ No newline at end of file
+    def _configure_tpdo(self):
+        with self._lock:
+            for motor_controller in self._motor_controllers:
+
+                try:
+                    motor_controller.can_node.tpdo.read()
+                except canopen.SdoCommunicationError:
+                    rospy.logwarn(f'[{rospy.get_name()}] Failed to read TPDO configuration.')
```

Comment:
This warning pops up every time the driver is started, even though everything is running as planned. Is there anything we can do about it? E.g. to make the first warning pop up after 2-3 read attempts.